tools/doc/html.js

```diff
@@ -23,51 +23,71 @@
+const visit = require('unist-util-visit');
+const markdown = require('remark-parse');
+const remark2rehype = require('remark-rehype');
+const raw = require('rehype-raw');
+const html = require('rehype-stringify');
 const path = require('path');
 const typeParser = require('./type-parser.js');
 
 module.exports = toHTML;
 
-// Make `marked` to not automatically insert id attributes in headings.
-const renderer = new marked.Renderer();
-renderer.heading = (text, level) => `<h${level}>${text}</h${level}>\n`;
-marked.setOptions({ renderer });
-
 const docPath = path.resolve(__dirname, '..', '..', 'doc');
 
+// Add class attributes to index navigation links.
+function navClasses() {
+  return walker;
+
+  function walker(node, file) {
+    if (node.type === 'element' && node.tagName === 'a') {
+      node.properties.class = 'nav-' +
+        node.properties.href.replace('.html', '').replace(/\W+/g, '-');
+    }
+
+    if (node.children) {
+      for (const child of node.children) walker(child, file);
+    }
+  }
+}
+
 const gtocPath = path.join(docPath, 'api', '_toc.md');
 const gtocMD = fs.readFileSync(gtocPath, 'utf8').replace(/^@\/\/.*$/gm, '');
-const gtocHTML = marked(gtocMD).replace(
-  /<a href="(.*?)"/g,
-  (all, href) => `<a class="nav-${href.replace('.html', '')
-                                      .replace(/\W+/g, '-')}" href="${href}"`
-);
+const gtocHTML = unified()
+  .use(markdown)
+  .use(remark2rehype, { allowDangerousHTML: true })
+  .use(raw)
+  .use(navClasses)
+  .use(html)
+  .processSync(gtocMD).toString();
 
 const templatePath = path.join(docPath, 'template.html');
 const template = fs.readFileSync(templatePath, 'utf8');
 
 function toHTML({ input, filename, nodeVersion, analytics }, cb) {
   filename = path.basename(filename, '.md');
 
-  const lexed = marked.lexer(input);
-
-  const firstHeading = lexed.find(({ type }) => type === 'heading');
-  const section = firstHeading ? firstHeading.text : 'Index';
-
-  preprocessText(lexed);
-  preprocessElements(lexed, filename);
-
-  // Generate the table of contents. This mutates the lexed contents in-place.
-  const toc = buildToc(lexed, filename);
+  const content = unified()
+    .use(markdown)
+    .use(firstHeader)
+    .use(preprocessText)
+    .use(preprocessElements, { filename: filename })
+    .use(buildToc, { filename: filename })
+    .use(remark2rehype, { allowDangerousHTML: true })
+    .use(raw)
+    .use(html)
+    .processSync(input);
 
   const id = filename.replace(/\W+/g, '-');
 
   let HTML = template.replace('__ID__', id)
                      .replace(/__FILENAME__/g, filename)
-                     .replace('__SECTION__', section)
+                     .replace('__SECTION__', content.section)
                      .replace(/__VERSION__/g, nodeVersion)
-                     .replace('__TOC__', toc)
+                     .replace('__TOC__', content.toc)
+                     .replace('__CONTENT__', content.toString())
```

Comment:
I am not sure if this is significant, but is it worth to make this change at the very end? We have some replacing below, and `content.toString()` is the biggest insertion here. Would the `__GTOC__`, `analytics` and `docCreated` insertions be faster if it is made after `__CONTENT__` replacement? Because this will make `.replace()` calls operate on a smaller string.